### ISSUE TITLE
Explicitly specify column data types for `load_csv_files`

### ISSUE THREAD

**Issue by ashiklom** [JGCRI/gcamdata#1072]
Although it generally does a good job, we probably shouldn't rely on `readr`'s automatic column type detection for data that we store inside the instead. Instead, we should explicitly pass the `col_types` argument, which can be either a single abbreviated character string (`"ccind"`, which means "character", "character", "integer", "number/double", "date") or a list of column specifications.

A convenient way to do this would be to store the column type information in the file header. This is the approach I took in `metar` (#916; see [example file](https://github.com/ashiklom/metar/blob/master/inst/examples/example1.csvy#L5-L20)), though this requires the file to be read in two stages: first, parse the header to get the column types; and second, read the file body with the indicated header types. An advantage of using a `metar`-style YAML header is that it can readily be adapted to store factor levels, which can be another potential defense against typos in columns. more generally, YAML is a very versatile format that can be used to store a wide range of data structures, and is widely used.

As an alternative, we could store column types as another tag within the current header. We may even be able to store factor levels for some columns as a list, e.g.

```
# ...
# ColumnTypes: integer, date, factor, number
# FactorLevels: , , ["apples", "oranges", "bananas"], 
id, date, fruit, value,
1, 2018-10-10, "apple", 5,
2, 2018-10-11, "orange", 6
...
```

**Comment by ashiklom** [comment]
The impetus for this is that `readr` 1.2.0 (November 2018) changed its parsing rules so that it never detects integer columns by default. This broke downstream code with strict type checking via `dplyr::if_else`.

**Comment by bpbond** [comment]
This would have the benefit that we could get rid of @rplzzz 's "protect_integer_column" routine (or whatever it's called).

I could just store the `ColumnTypes` as a `readr`-style short string, e.g. `# ColumnTypes: ccicddc`

**Comment by bpbond** [comment]
...so we just parse and stick directly into `read_csv`'s `col_types` argument.

**Comment by rplzzz** [comment]
> This would have the benefit that we could get rid of @rplzzz 's "protect_integer_column" routine (or whatever it's called).
>

I don't think this has anything to do with `protect_integer_cols`, which is there to deal with columns whose _names_ are integers, not columns whose _contents_ are integers.  The problem is that a column name that is an integer can in certain circumstances be interpreted as a column index, as explained in the documentation for that function:

https://github.com/JGCRI/gcamdata/blob/d2e0672dc8e1a0c7aca7e07af41a840776ca7d4d/R/pipeline-helpers.R#L236-L251

**Comment by bpbond** [comment]
@rplzzz I was thinking of `protect_float`.

**Comment by bpbond** [comment]
Before dealing with this (which is high priority, because we're now forcing users to downgrade their `readr` versions) it would be useful to merge #1070 or else I think we'll have lots of conflicts. Is anyone available to review that? @rplzzz @ashiklom @pralitp @kvcalvin

**Comment by pralitp** [comment]
Sure, I'll take that.  It looks straight forward enough.

**Comment by bpbond** [comment]
Thanks, Pralit.
